salt/file/xmpp/ejabberd_authentication.py

```diff
@@ -0,0 +1,170 @@
+alice@example.com and her password were "password", the file might look like:
+
+alice:example.com:$6$wntdG0RUDD9lt304$06Lf2Lrlqn7rgdTc3VL4lFjUOz3AadknDPEmrpWyFr5Jzkv9lQyyRz7mWYJ/ILnBRHfHbori.X4sR9B5DcKB60
+
+This script tries to avoid disclosing information via timing attacks as much as
+possible, so --max-passwords-per-user is designed so that (in theory) it should
+take the same amount of time to check a password regardless of how many
+passwords a user has configured.
+"""
+# Protocol: https://docs.ejabberd.im/developer/guide/#external
+#
+# When working on this script, keep these things in mind:
+#
+# It's fine if misconfiguration causes an exception at startup (but not after
+# startup) or if a protocol error causes an exception. However, untrusted input
+# (e.g., usernames, server names, or passwords) must not crash the script.
+#
+# Exception or log messages must not leak passwords or other sensitive
+# information, so they generally also shouldn't include any unknown information.
+# E.g., "parse error, unexpected bytes b'foo'" could risk leaking sensitive
+# information.
+
+import argparse
+import collections
+import crypt
+import enum
+import hmac
+import pathlib
+import struct
+import sys
+from typing import Collection, Mapping, Tuple
+
+# Map from (user, server) to that user's crypted passwords.
+_Config = Mapping[Tuple[bytes, bytes], Collection[str]]
+
+
+def _args():
+    parser = argparse.ArgumentParser(
+        description='External authentication helper for ejabberd.')
+    parser.add_argument(
+        '--config',
+        type=pathlib.Path,
+        required=True,
+        help='Authentication config file.',
+    )
+    parser.add_argument(
+        '--max-passwords-per-user',
+        default=25,
+        type=int,
+        help='Max number of passwords per user.',
+    )
+    return parser.parse_args()
+
+
+def _config(
+    config_path: pathlib.Path,
+    *,
+    max_passwords_per_user: int,
+) -> _Config:
+    raw_config = collections.defaultdict(list)
+    with config_path.open(mode='rb') as config_file:
+        for line in config_file:
+            user, server, crypted_password = line.rstrip(b'\n').split(b':')
+            raw_config[(user, server)].append(crypted_password.decode('utf-8'))
+    config = {}
+    for key, crypted_passwords in raw_config.items():
+        if len(crypted_passwords) > max_passwords_per_user:
+            raise ValueError(f'{key!r} has too many passwords')
+        config[key] = tuple(crypted_passwords[i % len(crypted_passwords)]
+                            for i in range(max_passwords_per_user))
+    return config
+
+
+class _Response(enum.IntEnum):
+    FAILURE = 0
+    SUCCESS = 1
+
+
+def _read() -> Tuple[bytes, bytes]:
+    length_bytes = sys.stdin.buffer.read(2)
+    if not length_bytes:
+        raise EOFError()
+    elif len(length_bytes) != 2:
+        raise ValueError(f'Expected 2 bytes, got {len(length_bytes)}')
+    (length,) = struct.unpack('>H', length_bytes)
```

Comment:
nit: ejabberd docs say "network byte order", so I think `!` is better:

```suggestion
    (length,) = struct.unpack('!H', length_bytes)
```

Reply:
Done.